[api-minor] Expanding divs to improve selection

jeremypress opened the pull request:
# Description

Continuing the work by @yurydelendik, I'm attempting to get his commit into shape to be included in the library. So far, I've converted his feature into an option that is set when defining the PDFViewerApplication, and have run some preliminary testing. 
# Performance Testing

So far I have tested 2 aspects of performance in Chrome. Memory/CPU usage, and the time taken to render text layers and set the text content, as previously logged by Yury. I chose 5 PDFs that vary in size, amount of text, and complexity to try to get a wide base of results. I've attached excel documents of my findings, and I'll summarize them here.

Here are the files I've used for testing so far: 
[pdfs.zip](https://github.com/mozilla/pdf.js/files/415937/pdfs.zip)
## CPU and Memory Testing

This was done with a modification of a script provided by Yury that logs CPU and Memory usage every 300 ms. Additionally, I wrote an apple script to simulate scrolling on the browser. First scrolling up, then down, then jumping to the last page and scrolling up. This allows me to replicate and time the results from the script. The scripts can be found here: https://github.com/jeremypress/browser-testing/tree/master. Below are some graphs comparing usage between the two branches. There seems to be a negligible difference in CPU and Memory, but further testing may determine otherwise. 
[chromeperformance.xlsx](https://github.com/mozilla/pdf.js/files/415936/chromeperformance.xlsx)

File 1: 5.8mb, 200 pages
<img width="729" alt="blueprint-cpumem" src="https://cloud.githubusercontent.com/assets/2474027/17633578/db71a236-6082-11e6-9353-9d92dc350791.png">

File 2: 4.2mb, 81 pages
<img width="584" alt="webdesigner-cpumem" src="https://cloud.githubusercontent.com/assets/2474027/17633580/e156b38a-6082-11e6-84c9-cf0024cd4def.png">

File 3: 620kb, 3 pages
<img width="627" alt="california-cpumem" src="https://cloud.githubusercontent.com/assets/2474027/17633582/e4b81424-6082-11e6-85de-5aa23a327a9c.png">
## Render Times

Data on render times was logged using console.time statements written by Yury. Here is where the performance difference between the two branches can be noted. On text heavy files, render times can be as much as twice a slow, which would correlate to around a half second increase in total rendering time spent. Also, in some cases the fairexpand branch requires a small amount of extra divs to be rendered. However, this does not really impact first impression, as most of the rendering is done as a user moves to a new page. A sample of the data is provided below, and the full file is attached. 
[render times.xlsx](https://github.com/mozilla/pdf.js/files/415934/render.times.xlsx)

File 1: 5.8mb, 200 pages
<img width="1503" alt="blueprint-render" src="https://cloud.githubusercontent.com/assets/2474027/17633588/e9afa672-6082-11e6-9287-f30645a30532.png">

File 2: 4.2mb, 81 pages
<img width="1504" alt="webdesigner-render" src="https://cloud.githubusercontent.com/assets/2474027/17633590/ecd9b40a-6082-11e6-8ff2-091964bfc0cb.png">

File 3: 620kb, 3 pages
<img width="1507" alt="california-render" src="https://cloud.githubusercontent.com/assets/2474027/17633591/efe3596c-6082-11e6-93ae-27c476c9e62a.png">
## Handling rtl and rotated/vertical text

Overall I've noticed that this enhancement handles rtl selection fine, and generally reduces flickering on rotated text. Although there are some instances of rotated text where fairexpand makes it harder or impossible to select, these are already problem areas on the master branch. Attached are some videos showing the differences, and screenshots of the text divs on a document with vertical text. 

Videos: https://cloud.box.com/s/1twc6zhcf1lswo2pe1vqzmrrjb6btxwy

Screenshots:
master branch:
<img width="993" alt="screen shot 2016-08-11 at 1 23 30 pm" src="https://cloud.githubusercontent.com/assets/2474027/17631353/a52e39c8-6078-11e6-87e1-6a10d8975dc8.png">

<img width="993" alt="screen shot 2016-08-11 at 1 26 10 pm" src="https://cloud.githubusercontent.com/assets/2474027/17631365/b12637ee-6078-11e6-8f86-20d6db9958a1.png">

fairexpand branch:
<img width="991" alt="screen shot 2016-08-11 at 1 24 34 pm" src="https://cloud.githubusercontent.com/assets/2474027/17631374/b70d7730-6078-11e6-89e2-e9fa6d6e3bd9.png">

<img width="992" alt="screen shot 2016-08-11 at 1 25 40 pm" src="https://cloud.githubusercontent.com/assets/2474027/17631375/b7252632-6078-11e6-800d-993b586b8a4b.png">
# Concerns
1. Chrome is the only browser I've tested in so far, what other data is desired in order to determine if the commit is performant enough?
2. Is the option implemented correctly according to the design of the project?
3. I'm failing a unit test regarding rendering PDF's in parallel. However, this also happens on my master branch, so it could be due to my testing environment. 
   
   Feedback is greatly appreciated so I can make improvements going forward.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/pdf.js/7539)

<!-- Reviewable:end -->

yurydelendik commented:
Overall looks good. Can you squash all commits into one (using `git rebase -i` command)? Also remove all instrumentation console.time/timeEnd/log operations. You can create addition commit with console instrumentation, which you can remove (via `git rebase -i`) before landing.

Also in separate commit modify test driver to get the extension setting from the manifest and apply that for the "text" tests and introduce couple of tests with this setting.

---

Reviewed 6 of 7 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[web/app.js, line 364 \[r2\]](https://reviewable.io:443/reviews/mozilla/pdf.js/7539#-KPJryYqxt1SRMBgf7AI:-KPJryYrCcRyN5U3-jfn:b-w1lrfn) ([raw file](https://github.com/mozilla/pdf.js/blob/f2a1093b44500e28560eb316fafcfc41e454da9f/web/app.js#L364)):*
> ```JavaScript
>         PDFJS.externalLinkTarget = value;
>       }),
> ```

unneeded change

---

*[web/text_layer_builder.js, line 25 \[r2\]](https://reviewable.io:443/reviews/mozilla/pdf.js/7539#-KPJrjwiZjXqNwQut2Eg:-KPJrjwiZjXqNwQut2Eh:b-lmh5o8) ([raw file](https://github.com/mozilla/pdf.js/blob/f2a1093b44500e28560eb316fafcfc41e454da9f/web/text_layer_builder.js#L25)):*
> ```JavaScript
>   return !NonWhitespaceRegexp.test(str);
> }
> ```

We don't need the code above.

---

*[web/text_layer_builder.js, line 91 \[r2\]](https://reviewable.io:443/reviews/mozilla/pdf.js/7539#-KPJrdJaP82cgTpxGhAp:-KPJrdJaP82cgTpxGhAq:bmhwqt9) ([raw file](https://github.com/mozilla/pdf.js/blob/94089960c04d7f59eb637d6dc63944115bbc539d/web/text_layer_builder.js#L91)):*
> ```JavaScript
>         this.textLayerRenderTask = null;
>       }
> ```

Unneeded change

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/pdf.js/7539#-:-KPJuKkiq60OUpm2s7Vr:bclwcu)*
<!-- Sent from Reviewable.io -->

Commit:
Continuing fairexpand #6663
1. Expanding divs to improve text selection. (Yury)
2. Adding enhanceTextSelection as an option.
3. Moving feature functionality from text_layer_builder.js to text_layer.js.
4. Added expandTextDivs method to only load expanded divs on first click, and only show on subsequent clicks

timvandermeij commented on web/text_layer_builder.js:
```diff
@@ -96,14 +106,15 @@ var TextLayerBuilder = (function TextLayerBuilderClosure() {
         this.updateMatches();
       }.bind(this), function (reason) {
-        // canceled or failed to render text layer -- skipping errors
+          // canceled or failed to render text layer -- skipping errors
```
Nit: revert this change as everything should be indented with two spaces

timvandermeij commented on web/app.js:
```diff
@@ -203,13 +204,16 @@ var PDFViewerApplication = {
 
     var container = appConfig.mainContainer;
     var viewer = appConfig.viewerContainer;
+    console.log('Enhanced Text Selection is set to:',
```
Nit: remove this `console.log`.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
Is this disabled by default to first land the code for more testing and enabling it later? If so, when this lands we should remember to file a follow-up issue to enable this mode by default.

yurydelendik commented:
TODOs:
  - [ ] decide if we want to enable this by default (with `//TODO` comments not-addressed`)
  - [ ] remove/disable bindMouse in the text_layer_builder.js depending on options
  - [ ] add "text" test(s) for enhanced mode to track regressions for it

Both TODO-comments (red-black tree and reference counting) decrease algorithmic complexity of the feature. However it looks like we are not dealing with large amount of text items on a "horizon" -- large amount of text items must be in single row or column in random order. I don't think it can be found in real world documents (?), so we can keep it as is for release.

yurydelendik commented:
Reviewed 1 of 3 files at r3.
Review status: 4 of 6 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/pdf.js/7539)*
<!-- Sent from Reviewable.io -->

Commit:
added text enhancement regression tests

timvandermeij commented on test/driver.js:
```diff
@@ -468,12 +472,13 @@ var Driver = (function DriverClosure() {
               var textLayerContext = textLayerCanvas.getContext('2d');
               textLayerContext.clearRect(0, 0,
                 textLayerCanvas.width, textLayerCanvas.height);
+              var enhanceText = task.id.includes('enhance') ? true : false;
```
`includes` returns a boolean, so we can simplify this line to `var enhanceText = task.id.includes('enhance');`

Snuffleupagus replied:
I'd would probably be much better overall if a `enhance` parameter was added (similar to the `annotations` flag in PR #6780).

timvandermeij replied:
Good, we had the same thought :)

I would personally prefer `enhance_text_selection` to make it clear what is being enhanced, much like the already used `enhanceTextSelection` variable in the code.

Snuffleupagus replied:
Wouldn't just `enhance` be enough though, since the type is already set to `text`?

timvandermeij replied:
Of couse, I forgot about that. Yes, `enhance` would be fine then.

timvandermeij commented on src/display/text_layer.js:
```diff
@@ -56,7 +56,12 @@ var renderTextLayer = (function renderTextLayerClosure() {
   }
 
-  function appendText(textDivs, viewport, geom, styles) {
+  function appendText(textDivs,
```
Convention for the PDF.js project is to not use a new line for every arguments as this list is taking up quite some space. We keep the arguments on one line and only add a second line if it does not fit into 80 characters anymore. In this case it would be preferred to have:

<pre>
function appendText(textDivs, viewport, geom, styles, bounds,
                                   enhanceTextSelection) {
  ...
}
</pre>

timvandermeij commented on web/text_layer_builder.js:
```diff
@@ -96,14 +99,15 @@ var TextLayerBuilder = (function TextLayerBuilderClosure() {
         this.updateMatches();
       }.bind(this), function (reason) {
-        // canceled or failed to render text layer -- skipping errors
+      // canceled or failed to render text layer -- skipping errors
```
Nit: undo this change, as the current situation (with the two spaces) is correct according to the style guide.

timvandermeij commented on web/pdf_page_view.js:
```diff
@@ -395,9 +399,11 @@ var PDFPageView = (function PDFPageViewClosure() {
-                                                                 this.viewport);
+        textLayer = this.textLayerFactory.
+          createTextLayerBuilder(textLayerDiv,
+                                 this.id - 1,
```
The note about keeping parameters on one line where possible applies here too.

timvandermeij commented on test/test_manifest.json:
```diff
@@ -331,12 +331,26 @@
        "lastPage": 4,
        "type": "text"
     },
+    {  "id": "taro-text-enhance",
```
Let's add a testcase that uses `enhance` for the `tracemonkey` file too. Then way we should have horizontal, vertical and rotated text covered, as well as our default testing file, the Tracemonkey paper.

timvandermeij commented:
This patch is looking much better already. Please address the comments above and squash the commits into one commit.

jeremypress commented:
Hey guys, thanks for all the feedback. I think I addressed everything, but let me know if I missed something.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -168,23 +197,306 @@ var renderTextLayer = (function renderTextLayerClosure() {
    */
-  function TextLayerRenderTask(textContent, container, viewport, textDivs) {
+  function TextLayerRenderTask(textContent, container, viewport, textDivs,
+                              enhanceTextSelection) {
```
Nit: please increase the indentation with one space on this line, so that the parameters line up; i.e.
```javascript
  function TextLayerRenderTask(textContent, container, viewport, textDivs,
                               enhanceTextSelection) {
```

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -213,10 +526,15 @@ var renderTextLayer = (function renderTextLayerClosure() {
       for (var i = 0, len = textItems.length; i < len; i++) {
-        appendText(textDivs, viewport, textItems[i], styles);
+        appendText(textDivs, viewport, textItems[i], styles, bounds,
+                     enhanceTextSelection);
```
Nit: please decrease the indentation with two spaces on this line, so that the parameters line up; i.e.
```javascript
        appendText(textDivs, viewport, textItems[i], styles, bounds,
                   enhanceTextSelection);
```

Snuffleupagus commented on test/driver.js:
```diff
@@ -64,7 +64,10 @@ var rasterizeTextLayer = (function rasterizeTextLayerClosure() {
+  function rasterizeTextLayer(ctx,
+                              viewport,
+                              textContent,
+                              enhanceTextSelection) {
```
Nit: please change this to:
```javascript
  function rasterizeTextLayer(ctx, viewport, textContent,
                              enhanceTextSelection) {
```

Snuffleupagus commented:
> [...] decide if we want to enable this by default (with //TODO comments not-addressed)

So that we don't forget, note that this PR still has the problem that was outlined in https://github.com/mozilla/pdf.js/pull/6663#issuecomment-158238124:
>This does have the side affect of always showing the text cursor no matter where you hover over the page (instead of just over text).

yurydelendik commented:
I'm thinking we can expand on mousedown capture phase and return to previous state at mouseup:

  - [ ] check if we can split text_layer.js logic into enable/disable expansion.

If not, we might need to keep it off and expose via default_preferences.js

jeremypress commented:
![fair expand on click](https://cloud.githubusercontent.com/assets/2474027/17790095/e9714136-6549-11e6-8a6b-eefa5257c55c.gif)

Here's a snippet of what the div expansion on click looks like. This should solve the cursor problem. I just need to cleanup the code (as well as make the other style fixes above) and turn this into an option so that we can still run the enhancement regression tests.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -112,9 +113,37 @@ var renderTextLayer = (function renderTextLayerClosure() {
   }
 
-  function render(task) {
+  function render(task, enhanceOnClick) {
```
Why is `enhanceOnClick` needed here, couldn't you just `tast._enhanceOnClick` inside the function instead?

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -213,19 +531,85 @@ var renderTextLayer = (function renderTextLayerClosure() {
-
       if (!timeout) { // Render right away
-        render(this);
+        render(this, enhanceOnClick);
```
Comment https://github.com/mozilla/pdf.js/pull/7539/commits/a02d66a1089586a29f1afb7e2338257aa0e73374#r75456925 applies both here, and below as well.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -213,19 +531,85 @@ var renderTextLayer = (function renderTextLayerClosure() {
       }
+    },
+
+    _bindMouse: function TextLayer_bindMouse(textDivs) {
```
To me, it seems *very inefficient* to add event listeners to *every* single `textDiv` like this.
In practice, this also means that the expansion will not happen if text-selection is started *outside* of a `textDiv`.

I'm wondering if wouldn't make more sense to register the events on the just the `textLayerDiv` instead? (This might require that a reference to the `textLayerDiv` is passed in to `renderTextLayer`.)

Snuffleupagus replied:
Also, I don't think that you need to pass `textDivs` to this method. You should be able to use `this._textDivs` instead.

jeremypress replied:
Agreed, I'll look into your suggestions and make the other fixes as well. Thanks for the comments!

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -213,19 +531,85 @@ var renderTextLayer = (function renderTextLayerClosure() {
+    },
+    _onMousedown: function TextLayer_onMousedown(textDivs) {
+      return function() {
+        for (var i in textDivs) {
```
Nit: Since you know that `textDivs` is an Array, it's probably more efficient to instead do:
```
for (var i = 0, ii = textDivs.length; i < ii; i++) {
```
to iterate through the elements, given that `for..in` will unnecessarily stringify the indexes.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -213,19 +531,85 @@ var renderTextLayer = (function renderTextLayerClosure() {
+      }
+    },
+    _onMousedown: function TextLayer_onMousedown(textDivs) {
+      return function() {
```
Nit: You can remove this `function`, and the `textDivs` parameter can be replaced with `this._textDivs`, if you instead register the event handler with e.g. `this._onMouseup.bind(this)`.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -182,9 +495,12 @@ var renderTextLayer = (function renderTextLayerClosure() {
    * @param {PageViewport} viewport
    * @param {Array} textDivs
+   * @param {boolean} enhanceTextSelection
+   * @param {boolean} enhanceOnClick
```
As far as I can tell, with the exception of the tests, there's no reason why you'd ever not want the "expand-on-click" behaviour in the viewer.
Hence I'm wondering if it wouldn't make more sense to invert this parameter to instead be `false` by default, and requiring the user (i.e. the test driver) to pass in something to *explicitly* disable the "expand-on-click" behaviour, e.g. with a parameter like `disableExpandOnClick`, `forceExpand`, or similar.

jeremypress commented:
I'm still working out the state of the flags. I think there should just be one, that determines whether the divs are expanded on click, or are always expanded for testing. This most recent update only attaches a listener to the textLayer, and adds a method to text_layer_builder to bind the mouse to it. Ideally, I can edit the logic to only render the expanded divs on click, but that is proving to be challenging because it means two renders happening at different times. Once this is figured out, there won't be a performance hit at all on impression, and the time should be negligible rendering one page of divs on first highlight only.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -226,7 +529,54 @@ var renderTextLayer = (function renderTextLayerClosure() {
+    },
+    _expandTextDivs: function TextLayer_expandTextDivs(expand) {
+      if (expand) {
+        for (var i = 0, ii = this._textDivs.length - 1; i < ii; i++) {
```
It looks like this loop will skip the *last* element in the array, I think it must be like this instead:
```javascript
for (var i = 0, ii = this._textDivs.length; i < ii; i++) {
```

jeremypress replied:
I'm skipping the last div because it's an end of content div. I don't think it has any of the actual text.

Snuffleupagus replied:
I might be misunderstanding, but are you sure about that?
Note that the `endOfContent` div is appended in `text_layer_builder.js` (and only to the DOM, and not `this.textDivs` there) and should thus not affect `this._textDivs` inside `text_layer.js`.

jeremypress replied:
I'll check again, I thought I saw errors coming from it but I could have been iterating on the wrong thing at that time.

Snuffleupagus commented:
I've played around with PR this a bit, and it seems to work quite well, nicely done!

It seems to me that we could perhaps improve e.g. the `mousedown/mouseup` event listeners (and a few other small things) a little bit, perhaps something along these lines: https://gist.github.com/Snuffleupagus/dee8be50da7962e5221180f029422d1d?

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -798,13 +802,15 @@ var PDFViewer = (function pdfViewer() {
         viewport: viewport,
-        findController: this.isInPresentationMode ? null : this.findController
+        findController: this.isInPresentationMode ? null : this.findController,
+        enhanceTextSelection: enhanceTextSelection
```
Nit: Since text selection isn't possible in Presentation Mode, it might be a bit more efficient to instead change this line to:
```avascript
enhanceTextSelection: this.isInPresentationMode ? false : enhanceTextSelection,
```

jeremypress replied:
cool! I updated the commit with the fixes you suggested. Also, I changed the logic around so that expand only gets called the first time you click on the text layer, and not on page render. I did this by turning bounds into a property of the object, and introducing the "expanded" flag to ensure expand only gets called once per page. As a result, I set the boolean to for enhanceTextSelection to true, since it only takes effect when a user highlights and otherwise has almost no effect. I can flip it back off if required.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -126,10 +155,11 @@ var renderTextLayer = (function renderTextLayerClosure() {
     if (textDivsLength > MAX_TEXT_DIVS_TO_RENDER) {
+      task._renderingDone = true;
       capability.resolve();
+
```
Nit: Unnecessary newline, please remove.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -182,17 +479,23 @@ var renderTextLayer = (function renderTextLayerClosure() {
     this._canceled = false;
     this._capability = createPromiseCapability();
     this._renderTimer = null;
+    this.bounds = [];
```
Nit: Since all the other "class" members use a leading underscore, let's be consistent and change this to `this._bounds` instead.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -182,17 +479,23 @@ var renderTextLayer = (function renderTextLayerClosure() {
     this._renderTimer = null;
+    this.bounds = [];
+    this._enhanceTextSelection = !!enhanceTextSelection;
+    this.expanded = false;
```
Similar to comment https://github.com/mozilla/pdf.js/pull/7539/files?diff=split#r75816230.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -226,7 +530,61 @@ var renderTextLayer = (function renderTextLayerClosure() {
+      }
+      if (!this.expanded) {
+        expand(this.bounds, this._viewport);
+        this.expanded = true;
```
Question: Provided that it would work correctly, would it make sense from a memory usage perspective to clear out `this._bounds` after `expand(...)` has been called (e.g. by doing `this._bounds.length = 0` here)?

jeremypress replied:
Good question, I'll look into that. The bounds shouldn't be needed again once the new values have been calculated.

Snuffleupagus commented on web/text_layer_builder.js:
```diff
@@ -314,30 +319,37 @@ var TextLayerBuilder = (function TextLayerBuilderClosure() {
+          }
+  //#endif
+          end.classList.add('active');
+          });
```
Please revert all the indentation/line-breaking changes made above.
(Also note that the preprocessor tags, i.e. lines starting with `//#` cannot be indented, since that actually breaks the preprocessor.)

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -226,7 +530,61 @@ var renderTextLayer = (function renderTextLayerClosure() {
         }, timeout);
       }
-    }
+    },
```
Nit: I know that we haven't always been consistent, but let's put a newline here between the functions to make the code a bit easier to read.

Snuffleupagus commented on web/app.js:
```diff
@@ -176,6 +176,7 @@ var PDFViewerApplication = {
   preferencePdfBugEnabled: false,
   preferenceShowPreviousViewOnLoad: true,
   preferenceDefaultZoomValue: '',
+  preferenceEnhanceTextSelection: true,
```
I'd assume that we want to make this controllable through a preference!?

If so, it needs to be added to https://github.com/mozilla/pdf.js/blob/master/web/default_preferences.json, and then fetched with the other ones as part of the `initialize` function, see https://github.com/mozilla/pdf.js/blob/master/web/app.js#L308.

timvandermeij replied:
I agree here, this should be a preference. As a side note, I do think that enabling it by default is fine. Before merging this PR we must make sure we test it properly, but if we do not enable it by default most likely we not get a feedback for it.

yurydelendik replied:
Let's enable it by default since it's not CPU hungry anymore for standard usage.

Snuffleupagus replied:
> By the way, if this is enabled by default, do we still need separate enhance tests?

I'd definitely say that we need the separate tests, since this PR doesn't expand the divs until you click the page, and without these specific tests we have no way of ensuring that we don't accidentally regress that mode.

jeremypress replied:
Are the tests I added sufficient? Or are more required?

timvandermeij replied:
I think the ones you have now are sufficient.

timvandermeij commented:
/botio-linux preview

yurydelendik commented on web/text_layer_builder.js:
```diff
@@ -314,30 +319,37 @@ var TextLayerBuilder = (function TextLayerBuilderClosure() {
       div.addEventListener('mousedown', function (e) {
+        if (self.enhanceTextSelection && self.textLayerRenderTask) {
+          self.textLayerRenderTask.expandTextDivs(true);
+        }
```
here and in mouseup, return from the function to avoid div-workaround logic.

Snuffleupagus replied:
Then I'd assume that we also want to change https://github.com/mozilla/pdf.js/blob/master/web/text_layer_builder.js#L67-L69 to something like:
```javascript
if (!this.enhanceTextSelection) {
  var endOfContent = document.createElement('div');
  endOfContent.className = 'endOfContent';
  this.textLayerDiv.appendChild(endOfContent);
}
```

jeremypress commented:
@Snuffleupagus I had trouble adding the option as a preference because the pdfViewer gets instantiated before the Preference initialize promise happens. I discussed this with @yurydelendik and I've added at as a constant in pdf_viewer.js.

yurydelendik commented:
Looks good to me. We may postpone optimizations such as red/black tree and memory usage for later. If we find a performance regression, we can disable that in the pdf_viewer.js file.

@timvandermeij or @Snuffleupagus can you glance code again and see if we can land it?

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -226,7 +528,63 @@ var renderTextLayer = (function renderTextLayerClosure() {
+        expand(this._bounds, this._viewport);
+        this._expanded = true;
+      }
+      this._bounds.length = 0;
```
Please move this into the `if(!this._expanded) { ... }` block just above, since it's not necessary to do it more than once.

Snuffleupagus commented on web/text_layer_builder.js:
```diff
@@ -35,6 +35,9 @@
  * @property {PDFFindController} findController
+ * @property {boolean} enhanceTextSelection - Option to turn on improved
+ * text selection.
+
```
Nit: Unnecessary newline.

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -45,6 +45,7 @@ var MAX_AUTO_SCALE = uiUtils.MAX_AUTO_SCALE;
 var CSS_UNITS = uiUtils.CSS_UNITS;
 var DEFAULT_SCALE = uiUtils.DEFAULT_SCALE;
 var DEFAULT_SCALE_VALUE = uiUtils.DEFAULT_SCALE_VALUE;
+var ENHANCE_TEXT_SELECTION = true;
```
Depending on the outcome of the questions I asked in https://github.com/mozilla/pdf.js/pull/7539#issuecomment-241995058, this line may or may not stay. But if the `ENHANCE_TEXT_SELECTION` is kept, I think that it should be placed with the `DEFAULT_CACHE_SIZE` "constant" below, since placing it amongst the imported symbols makes it a bit hard to find.

yurydelendik commented:
> @yurydelendik Opinions on the above?

I would keep enhanceTextSelection on core API side. I don't have opinion for the viewer API. Since it's new functionality, I recommend to keep components backward compatible, but in this case we may control selection enhancement state via constants. (Perhaps we can even remove those, but they provide some type of documentation for the literal boolean value.)

Snuffleupagus commented:
> I would keep enhanceTextSelection on core API side.

Agreed, which is why the example diff I included above only touched code in the `/web` folder.

> Since it's new functionality, I recommend to keep components backward compatible.

Agreed; but my point here is that with the current state of this PR, we're actually changing the default for the viewer components, with no way to opt-out short of modifying the code. Unless you directly set `PDFViewer.enhanceTextSelection = false` from outside of `PDFViewer`, but to me it seems a bit strange to have a setting that is controlled in a different way than the other existing options.

How about we instead do this:
```diff
diff --git a/web/app.js b/web/app.js
index f447830..3484ab7 100644
--- a/web/app.js
+++ b/web/app.js
@@ -101,6 +101,7 @@ var SCALE_SELECT_CONTAINER_PADDING = 8;
 var SCALE_SELECT_PADDING = 22;
 var PAGE_NUMBER_LOADING_INDICATOR = 'visiblePageIsLoading';
 var DISABLE_AUTO_FETCH_LOADING_BAR_TIMEOUT = 5000;
+var ENHANCE_TEXT_SELECTION = true;
 
 function configure(PDFJS) {
   PDFJS.imageResourcesPath = './images/';
@@ -209,7 +210,8 @@ var PDFViewerApplication = {
       eventBus: eventBus,
       renderingQueue: pdfRenderingQueue,
       linkService: pdfLinkService,
-      downloadManager: downloadManager
+      downloadManager: downloadManager,
+      enhanceTextSelection: ENHANCE_TEXT_SELECTION,
     });
     pdfRenderingQueue.setViewer(this.pdfViewer);
     pdfLinkService.setViewer(this.pdfViewer);
diff --git a/web/pdf_page_view.js b/web/pdf_page_view.js
index 7f09334..506f50a 100644
--- a/web/pdf_page_view.js
+++ b/web/pdf_page_view.js
@@ -50,8 +50,8 @@ var TEXT_LAYER_RENDER_DELAY = 200; // ms
  * @property {PDFRenderingQueue} renderingQueue - The rendering queue object.
  * @property {IPDFTextLayerFactory} textLayerFactory
  * @property {IPDFAnnotationLayerFactory} annotationLayerFactory
- * @property {boolean} enhanceTextSelection - Turns on text enhancement. Default
- * is false.
+ * @property {boolean} enhanceTextSelection - Turns on the text
+ *   selection enhancement. The default is `false`.
  */
 
 /**
@@ -81,7 +81,7 @@ var PDFPageView = (function PDFPageViewClosure() {
     this.viewport = defaultViewport;
     this.pdfPageRotate = defaultViewport.rotation;
     this.hasRestrictedScaling = false;
-    this.enhanceTextSelection = enhanceTextSelection;
+    this.enhanceTextSelection = enhanceTextSelection || false;
 
     this.eventBus = options.eventBus || domEvents.getGlobalEventBus();
     this.renderingQueue = renderingQueue;
diff --git a/web/pdf_viewer.js b/web/pdf_viewer.js
index c180c4f..7d2e73b 100644
--- a/web/pdf_viewer.js
+++ b/web/pdf_viewer.js
@@ -45,7 +45,6 @@ var MAX_AUTO_SCALE = uiUtils.MAX_AUTO_SCALE;
 var CSS_UNITS = uiUtils.CSS_UNITS;
 var DEFAULT_SCALE = uiUtils.DEFAULT_SCALE;
 var DEFAULT_SCALE_VALUE = uiUtils.DEFAULT_SCALE_VALUE;
-var ENHANCE_TEXT_SELECTION = true;
 var scrollIntoView = uiUtils.scrollIntoView;
 var watchScroll = uiUtils.watchScroll;
 var getVisibleElements = uiUtils.getVisibleElements;
@@ -77,6 +76,8 @@ var DEFAULT_CACHE_SIZE = 10;
  *   queue object.
  * @property {boolean} removePageBorders - (optional) Removes the border shadow
  *   around the pages. The default is false.
+ * @property {boolean} enhanceTextSelection - (optional) Enables the improved
+ *   text selection behaviour. The default is `false`.
  */
 
 /**
@@ -128,7 +129,7 @@ var PDFViewer = (function pdfViewer() {
     this.linkService = options.linkService || new SimpleLinkService();
     this.downloadManager = options.downloadManager || null;
     this.removePageBorders = options.removePageBorders || false;
-    this.enhanceTextSelection = ENHANCE_TEXT_SELECTION;
+    this.enhanceTextSelection = options.enhanceTextSelection || false;
 
     this.defaultRenderingQueue = !options.renderingQueue;
     if (this.defaultRenderingQueue) {
@@ -355,7 +356,7 @@ var PDFViewer = (function pdfViewer() {
             renderingQueue: this.renderingQueue,
             textLayerFactory: textLayerFactory,
             annotationLayerFactory: this,
-            enhanceTextSelection: this.enhanceTextSelection
+            enhanceTextSelection: this.enhanceTextSelection,
           });
           bindOnAfterAndBeforeDraw(pageView);
           this._pages.push(pageView);
```

yurydelendik commented:
As I understand, there is a problem in order of initialization of UI components and preferences. We probably need to change that to allow preferences to be initialized before we build an app, but that might be out-of-scope of this PR. A text layer builder factory is not easliy reachable via pdfViewer -- perhaps adding method or property at PDFViewer and TextLayerBuilderFactory might address that (but I bet this mode will stick with our demo viewer as default).

yurydelendik commented:
@jeremypress @Snuffleupagus the proposal at https://github.com/mozilla/pdf.js/pull/7539#issuecomment-242080291 sounds good to me.

Snuffleupagus commented:
/botio test

timvandermeij commented:
@jeremypress Just so you know, don't worry about the test failures. They are not related to this PR and are known to be intermittent.

speedplane commented:
I haven't been following this PR carefully, but you may want to make sure it performs decently when handling a PDF where each individual letter is its own textbox.  Here [is an example](http://www.music.mcgill.ca/~cmckay/papers/musictech/ISMIR_2006_Genre.pdf).

Snuffleupagus commented:
> [...] but you may want to make sure it performs decently when handling a PDF where each individual letter is its own textbox. Here is an example.

That probably isn't a good example any more, since that particular PDF file comes from issue #1045, which was fixed by PR #6590 :-)
A better example is probably the PDF file in issue #5808, which does seem to make the actual text-selection a bit sluggish unfortunately. (But note that PR #7387 seem to improve things quite a lot for that file.)

yurydelendik commented:
@jeremypress can you provide some stats for the PDFs in questions above? (if it's a regression we probably need to disable this by default for now until TODO-optimizations performed -- it will be nice to land this though with tests)

jeremypress commented:
@yurydelendik will do, tomorrow

Snuffleupagus commented:
Regarding the performance of documents with a large amount of `textDivs`, e.g. the examples provided in https://github.com/mozilla/pdf.js/pull/7539#issuecomment-242665191, I think that that there're two (somewhat) separate issues at play here:
 - First of all, there's the performance issues with the `expand` function (and the functions it calls). That function only runs once, and elsewhere in the PR there's some optimization ideas mentioned, so this part can probably be addressed.
 - Second of all, there's the performance issues with the `expandTextDivs` method that is called every time that the user clicks on a page. For documents with a very *large* number of `textDivs`, it seems that updating all of them on click is making the text-selection experience a bit sluggish. (This part of the problem might not be as easy to optimize as the first.)

Also besides the performance issues, when testing this PR on documents where every glyph gets its own `textDiv` (e.g. the PDF file in issue #7224), it seems that the `expand` function has some trouble with that kind of document since the text moves too far from its intended positions.

jeremypress commented:
@yurydelendik I've looked at the 2 files and here's what I found:

Memory/CPU: results were similar to previous testing, no significant differences. When it comes to selection, both this commit and master are similar as well.

Render Times: The introduction of the text div expansion adds between 150-100 ms on first click, and 60-100 ms on subsequent clicks. This is possibly the sluggish scrolling that @Snuffleupagus reported. I only noticed this very slightly on the initial click of the highlight, dragging it around still felt as performant to me as on master. (testing done with text layer not visible)

Highlight Fidelity: On these documents there isa drop in ability to highlight. I tested it with #7387 and it does significantly improve the situation. The problem can be seen in this [gif](http://giphy.com/gifs/l0MYQOJMi6NCORO9i), where the highlight starts where the expanded divs are set, and returns to their normal position on mouse up.

I don't know how many documents have these specific problems, but until #7387 lands I would consider this a regression. Additionally, with the potential optimization to the expand method, I see the fidelity as more of an issue than the performance

The last two sheets are the new files
[render times.xlsx](https://github.com/mozilla/pdf.js/files/445619/render.times.xlsx)

Performance results for the 2 new files
[results.zip](https://github.com/mozilla/pdf.js/files/445620/results.zip)

Snuffleupagus commented:
> I don't know how many documents have these specific problems, but until #7387 lands I would consider this a regression.

I just want to note here that even though PR #7387 improves *some* cases quite a lot (e.g. issue #5808), it unfortunately does nothing for e.g. issue #7224.

timvandermeij commented on src/display/text_layer.js:
```diff
@@ -126,10 +155,10 @@ var renderTextLayer = (function renderTextLayerClosure() {
       capability.resolve();
       return;
     }
-
```
Please undo this change.

timvandermeij commented on src/display/text_layer.js:
```diff
@@ -213,10 +514,11 @@ var renderTextLayer = (function renderTextLayerClosure() {
+        appendText(textDivs, viewport, textItems[i], styles, this._bounds,
+                   enhanceTextSelection);
       }
-
```
Please undo this change.

timvandermeij commented on src/display/text_layer.js:
```diff
@@ -213,10 +514,11 @@ var renderTextLayer = (function renderTextLayerClosure() {
       var textDivs = this._textDivs;
       var viewport = this._viewport;
+      var enhanceTextSelection = this._enhanceTextSelection;
       for (var i = 0, len = textItems.length; i < len; i++) {
```
Add an additional newline before the `for`.

timvandermeij commented on web/app.js:
```diff
@@ -209,7 +210,9 @@ var PDFViewerApplication = {
-      downloadManager: downloadManager
+      downloadManager: downloadManager,
+      enhanceTextSelection: ENHANCE_TEXT_SELECTION,
+
```
Please remove this newline.

timvandermeij commented on web/pdf_page_view.js:
```diff
@@ -69,6 +71,7 @@ var PDFPageView = (function PDFPageViewClosure() {
     var renderingQueue = options.renderingQueue;
     var textLayerFactory = options.textLayerFactory;
     var annotationLayerFactory = options.annotationLayerFactory;
+    var enhanceTextSelection = options.enhanceTextSelection || false;
```
The intention is to enable this by default (see previous comments), so let's change this to `true` (and in the comment above).

Snuffleupagus replied:
Given the performance/correctness issues currently present, I don't think that we want to enable this by default at this point (so `var ENHANCE_TEXT_SELECTION = false;` should be set in app.js).

More to the point though, please note that we agreed to https://github.com/mozilla/pdf.js/pull/7539#issuecomment-242080291, to avoid changing the defaults for the viewer components. Hence this comment is not applicable.

timvandermeij replied:
Agreed. We can also land this disabled and file a follow-up issue for the optimizations.

timvandermeij commented on web/pdf_viewer.js:
```diff
@@ -127,6 +129,8 @@ var PDFViewer = (function pdfViewer() {
     this.downloadManager = options.downloadManager || null;
     this.removePageBorders = options.removePageBorders || false;
+    this.enhanceTextSelection = options.enhanceTextSelection || false;
+
```
Please remove this newline.

timvandermeij commented on web/pdf_viewer.js:
```diff
@@ -127,6 +129,8 @@ var PDFViewer = (function pdfViewer() {
     this.linkService = options.linkService || new SimpleLinkService();
     this.downloadManager = options.downloadManager || null;
     this.removePageBorders = options.removePageBorders || false;
+    this.enhanceTextSelection = options.enhanceTextSelection || false;
```
Same here, let's make this `true` (and in the comment above).

Snuffleupagus replied:
https://github.com/mozilla/pdf.js/pull/7539#discussion_r76987913 applies here too.

timvandermeij commented:
Pull request #7387 landed. I have added some more review comments above. After addressing these, we'll build a preview and run the tests to see if we can land this.

jeremypress commented:
@timvandermeij I made the fixes, but left `|| false` in the last two per @Snuffleupagus 's comment. Very excited for this to land!

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

timvandermeij commented:
/botio makeref

timvandermeij commented:
Nice work! Thank you @yurydelendik for the initial patch and @jeremypress for finalizing the patch!

speedplane commented:
Definitely awesome work everyone! Can we get a sample build made so we can easily play around with it?

tonyjin commented:
!!!! this is the best

timvandermeij commented:
For now you have to set `enhanceTextSelection` to `true` to work with this mode. Now we have PR #7586 that adds it as a viewer preference so it can be toggled more easily.
